src/main/java/org/prebid/server/bidder/adgeneration/AdgenerationBidder.java

```diff
@@ -94,7 +94,7 @@ private ExtImpAdgeneration parseAndValidateImpExt(Imp imp) {
         return extImpAdgeneration;
     }
 
-    private String getUri(String adSize, String id, String currency, Site site, Source source) {
+    private String getUri(String adSize, String id, String currency, final BidRequest bidRequest) {
```

Comment:
Rewrite method like this:
```
private String getUri(String adSize, String id, String currency, final BidRequest bidRequest) {
    final URIBuilder uriBuilder;
    try {
        uriBuilder = new URIBuilder(endpointUrl);
    } catch (URISyntaxException e) {
        throw new PreBidException(String.format("Invalid url: %s, error: %s", endpointUrl, e.getMessage()));
    }

    uriBuilder
            .addParameter("posall", "SSPLOC")
            .addParameter("id", id)
            .addParameter("hb", "true")
            .addParameter("t", "json3")
            .addParameter("currency", currency)
            .addParameter("sdkname", "prebidserver")
            .addParameter("adapterver", VERSION);

    addUriParameterIfNotEmpty(uriBuilder, "sizes", adSize);
    addUriParameterIfNotEmpty(uriBuilder, "tp", ObjectUtil.getIfNotNull(bidRequest.getSite(), Site::getPage));
    addUriParameterIfNotEmpty(uriBuilder, "appbundle", ObjectUtil.getIfNotNull(bidRequest.getApp(), App::getBundle));
    addUriParameterIfNotEmpty(uriBuilder, "appname", ObjectUtil.getIfNotNull(bidRequest.getApp(), App::getName));
    addUriParameterIfNotEmpty(
            uriBuilder, "transactionid", ObjectUtil.getIfNotNull(bidRequest.getSource(), Source::getTid));

    final Device device = bidRequest.getDevice();
    final String deviceOs = device != null ? device.getOs() : null;
    if ("android".equals(deviceOs)) {
        uriBuilder.addParameter("sdktype", "1");
        addUriParameterIfNotEmpty(uriBuilder, "advertising_id", device.getIfa());
    } else if ("ios".equals(deviceOs)) {
        uriBuilder.addParameter("sdktype", "2");
        addUriParameterIfNotEmpty(uriBuilder, "idfa", device.getIfa());
    } else {
        uriBuilder.addParameter("sdktype", "0");
    }

    return uriBuilder.toString();
}

private static void addUriParameterIfNotEmpty(URIBuilder uriBuilder, String parameter, String value) {
    if (StringUtils.isNotEmpty(value)) {
        uriBuilder.addParameter(parameter, value);
    }
}
```

Reply:
In `go` each parameter goes in a chain, it maybe need to make the same. May effect on URI sequence or correct me.

Reply:
It does not matter what order the query parameters are in.